### PR TITLE
Reduce applicability requirements

### PR DESCRIPTION
**Pull request by dsavenko**
Hi,

We have a project type, which does not extend AbstractProject, but directly extends Job and implements Queue.Task (we have certain reasons to not extend AbstractProject). We would like to have the Throttle plugin working with it, but it's tied to AbstractProject. 

This pull request unties it from AbstractProject class and makes so, that the Throttle plugin works with any class which extends Job and implements Queue.Task. 

This is just a technical patch, no new features added.

### CODE REVIEW

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java [hunk 1]
```diff
@@ -24,20 +30,11 @@
...
-import org.kohsuke.stapler.DataBoundConstructor;
-import org.kohsuke.stapler.QueryParameter;
-import org.kohsuke.stapler.StaplerRequest;
-
-public class ThrottleJobProperty extends JobProperty<AbstractProject<?,?>> {
+public class ThrottleJobProperty extends JobProperty<Job<?,?>> {
     // Moving category to categories, to support, well, multiple categories per job.
     @Deprecated transient String category;
-    
+
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java]: Please avoid such formatting changes to simplify merges

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java [hunk 2]
```diff
@@ -24,20 +30,11 @@
...
-import jenkins.model.Jenkins;
 
-import net.sf.json.JSONObject;
-
-import org.kohsuke.stapler.DataBoundConstructor;
-import org.kohsuke.stapler.QueryParameter;
-import org.kohsuke.stapler.StaplerRequest;
-
-public class ThrottleJobProperty extends JobProperty<AbstractProject<?,?>> {
+public class ThrottleJobProperty extends JobProperty<Job<?,?>> {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java]: May be a binary compatibility issue, I need to check it

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java [hunk 3]
```diff
@@ -180,10 +180,10 @@ else if (tjp.getThrottleOption().equals("category")) {
 
     @CheckForNull
     private ThrottleJobProperty getThrottleJobProperty(Task task) {
-        if (task instanceof AbstractProject) {
-            AbstractProject<?,?> p = (AbstractProject<?,?>) task;
+        if (task instanceof Job) {
+            Job<?,?> p = (Job<?,?>) task;
             if (task instanceof MatrixConfiguration) {
-                p = (AbstractProject<?,?>)((MatrixConfiguration)task).getParent();
+                p = (Job<?,?>)((MatrixConfiguration)task).getParent();
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java]: BTW class cast to Job should not be required here

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java [hunk 4]
```diff
@@ -228,7 +228,7 @@ private int buildsOfProjectOnAllNodes(Task task) {
     private int buildsOnExecutor(Task task, Executor exec) {
         int runCount = 0;
         if (exec.getCurrentExecutable() != null
-            && exec.getCurrentExecutable().getParent() == task) {
+            && task.equals(exec.getCurrentExecutable().getParent())) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java]: This replacement is suspicious. Why did you need it?

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleJobPropertyTest.java [hunk 5]
```diff
@@ -1,8 +1,7 @@
 package hudson.plugins.throttleconcurrents;
 
-import hudson.model.AbstractProject;
-import hudson.model.FreeStyleProject;
-import hudson.model.Job;
+import hudson.model.*;
```
Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleJobPropertyTest.java]: It's a bad practice in the case of the non-test code

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java (hunk 1 above)
Comment (dsavenko) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java]: Sure, I will get rid of them. Actually, I removed them manually before commit, but IDE put them again :)

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java (hunk 4 above)
Comment (dsavenko) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java]: A couple of reasons:
1. Our job class has equals/hashCode overriden
2. In Jenkins core, they use equals for similar purposes, so I assume, that it is OK. E.g. here's a chunk of code from the hudson.model.Queue class:

    public boolean cancel(Task p) {
        lock.lock();
        try { try {
            LOGGER.log(Level.FINE, "Cancelling {0}", p);
            for (WaitingItem item : waitingList) {
                if (item.task.equals(p)) {
                    return item.cancel(this);
                }
            }
            // use bitwise-OR to make sure that both branches get evaluated all the time
            return blockedProjects.cancel(p) != null | buildables.cancel(p) != null;
        } finally { updateSnapshot(); } } finally {
            lock.unlock();
        }
    }

3. It's generally considered a good practice to use equals() instead of reference equality checks.

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleJobPropertyTest.java (hunk 5 above)
Comment (dsavenko) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleJobPropertyTest.java]: Sure, I will change it.

**Comment by jglick** [comment]
You probably want to implement `ParameterizedJobMixIn.ParameterizedJob` and have TCB look for that instead.

**Comment by dsavenko** [comment]
Good point, I haven't thought about it. Will provide the update soon.

**Comment by dsavenko** [comment]
@jglick, hmm... On second thought, ParameterizedJobMixIn is there since 1.556, while TCB is built against 1.424. Also, JobProperty's generic parameter must extend Job<?, ?>, so I'm bound to extend the ThrottleJobProperty class from JobProperty<Job<?,?>> (can't use ParameterizedJobMixIn.ParameterizedJob, as it doesn't extend Job).

**Comment by oleg-nenashev** [comment]
Personally I don't like the <code>SometihgMixIn</code> approach much, because it mostly workarounds Java's inheritance mechanism. It does not mean I can propose a better solution...

<code>ParameterizedJobMixIn.ParameterizedJob</code> is also not guaranteed to be used by all Job implementations, so I'd prefer the current approach.

In any case, the core dependency update may be reasonable. Versions 1.535...1.609.2 do not behave reliably, so probably bumping to 1.609.2 is the best option. The plugin needs some redesign, so please do not consider the version dependency as a blocker.

**Comment by dsavenko** [comment]
So, what should I do next?

**Comment by abayer** [comment]
Dependency update is perfectly fine, yeah.

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java (hunk 4 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java]: Agreed

**Comment by abayer** [comment]
fwiw, tested this and it works fine throttling groups of Workflow jobs - not being used within the Workflow script, though. That's more work, which I'll get diving into once this is in.

**Comment by oleg-nenashev** [comment]
:+1: for the current implementation

**Comment by abayer** [comment]
Rightie-o, then I'll merge it.

**Comment by dsavenko** [comment]
Thanks for merging!
